### PR TITLE
bluestore: Elastic Shared Blobs 1 - mixed preparation

### PR DESCRIPTION
**Pull request by aclamk**
This is part 1/4 of ESB work.
It contains fixes and improvements that are useful to BlueStore but not necessary, but are necessary to the ESB.
#51439 
#51440 
#51441 
#51442

## Contribution Guidelines
- To sign and title your commits, please refer to [Submitting Patches to Ceph](https://github.com/ceph/ceph/blob/main/SubmittingPatches.rst).

- If you are submitting a fix for a stable branch (e.g. "pacific"), please refer to [Submitting Patches to Ceph - Backports](https://github.com/ceph/ceph/blob/master/SubmittingPatches-backports.rst) for the proper workflow.

## Checklist
- Tracker (select at least one)
  - [ ] References tracker ticket
  - [ ] Very recent bug; references commit where it was introduced
  - [x] New feature (ticket optional)
  - [ ] Doc update (no ticket needed)
  - [ ] Code cleanup (no ticket needed)
- Component impact
  - [ ] Affects [Dashboard](https://tracker.ceph.com/projects/dashboard/issues/new), opened tracker ticket
  - [ ] Affects [Orchestrator](https://tracker.ceph.com/projects/orchestrator/issues/new), opened tracker ticket
  - [x] No impact that needs to be tracked
- Documentation (select at least one)
  - [ ] Updates relevant documentation
  - [x] No doc update is appropriate
- Tests (select at least one)
  - [ ] Includes [unit test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/tests-unit-tests/)
  - [ ] Includes [integration test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/testing_integration_tests/)
  - [ ] Includes bug reproducer
  - [x] No tests

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard cephadm`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`
- `jenkins test windows`
</details>

### CODE REVIEW

**Comment by aclamk** [comment]
jenkins test make check

File: src/os/bluestore/BlueStore.cc [hunk 1]
```diff
@@ -8083,12 +8083,15 @@ void BlueStore::_fsck_foreach_shared_blob(
...
-	  cb(c->cid, oid, sbid, b);
+	  if (cb(c->cid, oid, sbid, b) == false) {
+	    goto stop_iterating;
+	  }
 	  passed_sbs.emplace(e.blob);
 	}
       } // for ... extent_map
     } // for ... it->valid
   } //if (it(PREFIX_OBJ))
+ stop_iterating:;
```
Comment (rzarzynski) [inline src/os/bluestore/BlueStore.cc:8118]: Is there any reason in favor of `goto` than `return` (with a comment maybe)?

File: src/os/bluestore/BlueStore.cc [hunk 2]
```diff
@@ -9478,6 +9483,29 @@ int BlueStore::_fsck_on_open(BlueStore::FSCKDepth depth, bool repair)
...
+      _fsck_foreach_shared_blob( [&](coll_t cid,
+				     ghobject_t oid,
+				     uint64_t sbid,
+				     const bluestore_blob_t& b) {
+	for (auto& p : b.get_extents()) {
+	  if (p.is_valid() &&
+	      !sb_ref_counts.test_all_zero_range(sbid,
+						 p.offset,
+						 p.length)) {
+	    derr << "fsck possibly broken shared blob found for col:" << cid
```
[the remark this answers is not in the document]
Comment (aclamk) [reply src/os/bluestore/BlueStore.cc:9521]: done.

File: src/os/bluestore/BlueStore.cc (hunk 1 above)
Comment (aclamk) [reply src/os/bluestore/BlueStore.cc:8118]: done

**Comment by rzarzynski** [comment]
Merged as a part of https://github.com/ceph/ceph/pull/53178.

**Pull request closed without merging**